core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java

```diff
@@ -86,28 +95,44 @@ public String getName()
             return immediateVoidFuture();
         }
 
-        Optional<TableHandle> tableHandle = metadata.getTableHandle(session, tableName);
-        if (tableHandle.isEmpty()) {
+        RedirectionAwareTableHandle redirectionAwareTableHandle = metadata.getRedirectionAwareTableHandle(session, tableName);
+        if (redirectionAwareTableHandle.getTableHandle().isEmpty()) {
             if (!statement.isExists()) {
                 throw semanticException(TABLE_NOT_FOUND, statement, "Table '%s' does not exist", tableName);
             }
             return immediateVoidFuture();
         }
-
-        QualifiedObjectName target = createQualifiedObjectName(session, statement, statement.getTarget());
+        TableHandle tableHandle = redirectionAwareTableHandle.getTableHandle().get();
+        QualifiedObjectName source = redirectionAwareTableHandle.getRedirectedTableName().orElse(createQualifiedObjectName(session, statement, statement.getSource()));
+        QualifiedObjectName target = createTargetQualifiedObjectName(source, statement.getTarget());
         if (metadata.getCatalogHandle(session, target.getCatalogName()).isEmpty()) {
             throw semanticException(CATALOG_NOT_FOUND, statement, "Target catalog '%s' does not exist", target.getCatalogName());
         }
         if (metadata.getTableHandle(session, target).isPresent()) {
             throw semanticException(TABLE_ALREADY_EXISTS, statement, "Target table '%s' already exists", target);
         }
-        if (!tableName.getCatalogName().equals(target.getCatalogName())) {
+        if (!tableHandle.getCatalogName().getCatalogName().equals(target.getCatalogName())) {
             throw semanticException(NOT_SUPPORTED, statement, "Table rename across catalogs is not supported");
         }
-        accessControl.checkCanRenameTable(session.toSecurityContext(), tableName, target);
+        accessControl.checkCanRenameTable(session.toSecurityContext(), tableName, source);
```

Comment:
why target -> source ?

Reply:
do we have test coverage for this?

Reply:
We have a few tests:
- io.trino.security.TestAccessControl#testNonQueryAccessControl -> is using a rather naive io.trino.testing.TestingAccessControlManager behind the scenes (and no `systemAccessControls`)
- io.trino.testing.BaseConnectorTest#testRenameTable 

However I haven't found a test interacting with a fine grained system access control implementation like https://trino.io/docs/current/security/file-system-access-control.html

Reply:
Did any test fail when you had `target` here?

Reply:
No. There were no failures as far as I can recall.
https://github.com/trinodb/trino/actions/runs/1922889046

Reply:
Can you please add a test? Orthogonal to this PR

Reply:
actually…. there should be two tests: we should have an access control test for rename; and we should have an access control test for rename with redirects; we should verify which table AC is asked about

Reply:
Created https://github.com/trinodb/trino/pull/11488 as a scaffolding for dealing with access control tests for redirections. I'll be adding support for `RENAME TABLE` there once it reaches `master`.